File: Operation.java
public class Operation {
    private String firstOperand;
    private String secondOperand;
    private int digits;
    private char[] sum;
    private boolean overflow;
    private boolean signed;

    public Operation(StringBuilder firstOperand, StringBuilder secondOperand, int digits) {
        this.firstOperand   = firstOperand.toString();
        this.secondOperand  = secondOperand.toString();
        this.digits         = digits + 1;                   // binary digits + dot
        this.sum            = new char[this.digits + 1];    // (binary digits + dot) + overflow
        this.overflow       = false;
        this.signed         = false;

        for (int x = 0; x < this.digits + 1; x++) { sum[x] = ' '; }
    }

    public Operation(String firstOperand, String secondOperand, int digits) {
        this.firstOperand   = firstOperand;
        this.secondOperand  = secondOperand;
        this.digits         = digits + 1;                   // binary digits + dot
        this.sum            = new char[this.digits + 1];    // (binary digits + dot) + overflow
        this.overflow       = false;
        this.signed         = false;

        for (int x = 0; x < this.digits + 1; x++) { sum[x] = ' '; }
    }

    public char[] performComplement(String negative) {
        char[] positive = new char[this.digits];
        boolean firstOne = false;

        for (int x = 0; x < this.digits; x++) { positive[x] = ' '; }

        for (int x = this.digits - 1; x >= 0; x--) {
            if (negative.charAt(x + 1) == '1' && !firstOne) {
                positive[x] = '1';
                firstOne = true;
            } 
            else if (negative.charAt(x + 1) == '0' && !firstOne) { positive[x] = '0'; } 
            else if (negative.charAt(x + 1) == '1' && firstOne) { positive[x] = '0'; } 
            else if (negative.charAt(x + 1) == '0' && firstOne) { positive[x] = '1'; } 
            else { positive[x] = '.'; }
        }

        return positive;
    }

    public void performAddition() {
        if (this.firstOperand.charAt(0) == '-') {
            firstOperand = String.valueOf(performComplement(firstOperand));
            this.signed = true;
        } else if (this.secondOperand.charAt(0) == '-') {
            secondOperand = String.valueOf(performComplement(secondOperand));
            this.signed = true;
        }

        for (int x = this.digits - 1; x >= 0; x--) {
            if (this.firstOperand.charAt(x) == '1' && this.secondOperand.charAt(x) == '1' && this.overflow) {
                this.sum[x + 1] = '1';
                this.overflow   = true;
            } else if (this.firstOperand.charAt(x) == '1' && this.secondOperand.charAt(x) == '1' && !this.overflow) {
                this.sum[x + 1] = '0';
                this.overflow   = true;
            } else if (this.firstOperand.charAt(x) == '1' && this.secondOperand.charAt(x) == '0' && this.overflow
                    || this.firstOperand.charAt(x) == '0' && this.secondOperand.charAt(x) == '1' && this.overflow) {
                this.sum[x + 1] = '0';
                this.overflow   = true;
            } else if (this.firstOperand.charAt(x) == '1' && this.secondOperand.charAt(x) == '0' && !this.overflow
                    || this.firstOperand.charAt(x) == '0' && this.secondOperand.charAt(x) == '1' && !this.overflow) {
                this.sum[x + 1] = '1';
                this.overflow   = false;
            } else if (this.firstOperand.charAt(x) == '0' && this.secondOperand.charAt(x) == '0' && this.overflow) {
                this.sum[x + 1] = '1';
                this.overflow   = false;
            } else if (this.firstOperand.charAt(x) == '0' && this.secondOperand.charAt(x) == '0' && !this.overflow) {
                this.sum[x + 1] = '0';
                this.overflow   = false;
            } else { this.sum[x + 1] = '.'; }
        }

        if (overflow && !this.signed) { sum[0] = '1'; }
    }

    public char[] getSum(){ return sum; }

    public static void main(String[] args) {
        Operation operationW = new Operation("1.0100", "0.0101", 5);                // Without GRS
        operationW.performAddition();
        System.out.println(operationW.getSum());

        Operation operationGRS = new Operation("1.011111001", "0.010011111", 10);   // With GRS, add 3 digits for GRS
        operationGRS.performAddition();
        System.out.println(operationGRS.getSum());

        Operation operationN = new Operation("1.000", "-0.111", 4);                 // Negative, needs more checking
        operationN.performAddition();
        System.out.println(operationN.getSum());
    }
}

File: Operation.py
class Operation(object) :
    firstOperand = None
    secondOperand = None
    digits = 0
    sum = None
    overflow = False
    signed = False
    def __init__(self, firstOperand,  secondOperand,  grs,  digits) :
        self.firstOperand = firstOperand
        self.secondOperand = secondOperand
        if (not grs) :
            self.digits = digits + 1
        else :
            self.digits = (digits + 3) + 1
        self.sum = [' '] * (self.digits + 1)
        self.overflow = False
        self.signed = False
        x = 0
        while (x < self.digits + 1) :
            self.sum[x] = ' '
            x += 1
    def perform_addition(self) :
        if (self.firstOperand[0] == '-') :
            self.firstOperand = "".join(self.perform_complement(self.firstOperand))
            self.signed = True
        elif(self.secondOperand[0] == '-') :
            self.secondOperand = "".join(self.perform_complement(self.secondOperand))
            self.signed = True
        x = self.digits - 1
        while (x >= 0) :
            if (self.firstOperand[x] == '1' and self.secondOperand[x] == '1' and self.overflow) :
                self.sum[x + 1] = '1'
                self.overflow = True
            elif(self.firstOperand[x] == '1' and self.secondOperand[x] == '1' and not self.overflow) :
                self.sum[x + 1] = '0'
                self.overflow = True
            elif(self.firstOperand[x] == '1' and self.secondOperand[x] == '0' and self.overflow or self.firstOperand[x] == '0' and self.secondOperand[x] == '1' and self.overflow) :
                self.sum[x + 1] = '0'
                self.overflow = True
            elif(self.firstOperand[x] == '1' and self.secondOperand[x] == '0' and not self.overflow or self.firstOperand[x] == '0' and self.secondOperand[x] == '1' and not self.overflow) :
                self.sum[x + 1] = '1'
                self.overflow = False
            elif(self.firstOperand[x] == '0' and self.secondOperand[x] == '0' and self.overflow) :
                self.sum[x + 1] = '1'
                self.overflow = False
            elif(self.firstOperand[x] == '0' and self.secondOperand[x] == '0' and not self.overflow) :
                self.sum[x + 1] = '0'
                self.overflow = False
            else :
                self.sum[x + 1] = '.'
            x -= 1
        if (self.overflow and not self.signed) :
            self.sum[0] = '1'
    def  get_sum(self) :
        return "".join(self.sum)
    def  perform_complement(self, negative) :
        positive = [' '] * (self.digits)
        firstOne = False
        x = 0
        while (x < self.digits) :
            positive[x] = ' '
            x += 1
        x = self.digits - 1
        while (x >= 0) :
            if (negative[x + 1] == '1' and not firstOne) :
                positive[x] = '1'
                firstOne = True
            elif(negative[x + 1] == '0' and not firstOne) :
                positive[x] = '0'
            elif(negative[x + 1] == '1' and firstOne) :
                positive[x] = '0'
            elif(negative[x + 1] == '0' and firstOne) :
                positive[x] = '1'
            else :
                positive[x] = '.'
            x -= 1
        return positive